socs/snmp.py

```diff
@@ -0,0 +1,120 @@
+import txaio
+
+from pysnmp.hlapi.twisted import getCmd, SnmpEngine, CommunityData, UdpTransportTarget,\
+                                 ContextData, ObjectType, ObjectIdentity
+
+# For logging
+txaio.use_twisted()
+
+
+class SNMPTwister:
+    """Helper class for handling SNMP communication with twisted.
+
+    More information can be found in the pySNMP documentation. The
+    `SNMP Operations`_ page is particularly helpful for understanding the setup
+    of this object.
+
+    Note: This helper currently only supports SNMPv1.
+
+    .. _SNMP Operations: http://snmplabs.com/pysnmp/docs/pysnmp-hlapi-tutorial.html
+
+    Parameters
+    ----------
+    address : str
+        Address of the SNMP Agent to send GET/SET requests to
+    port : int
+        Associated port for SNMP communication. Default is 161
+
+    Attributes
+    ----------
+    snmp_engine : pysnmp.entity.engine.SnmpEngine
+        PySNMP engine
+    address : str
+        Address of the SNMP Agent to send GET/SET requests to
+    udp_transport : pysnmp.hlapi.twisted.transport.UdpTransportTarget
+        UDP transport for UDP over IPv4
+    log : txaio.tx.Logger
+        txaio logger object
+
+    """
+
+    def __init__(self, address, port=161):
+        self.snmp_engine = SnmpEngine()
+        self.address = address
+        self.udp_transport = UdpTransportTarget((address, port))
+        self.log = txaio.make_logger()
+
+    def _success(self, args):
+        """Success callback for getCmd.
+
+        Taken from Twisted example for SNMPv1 from pySNMP documentation:
+        http://snmplabs.com/pysnmp/examples/hlapi/twisted/contents.html
+
+        Returns
+        -------
+        list
+            A sequence of ObjectType class instances representing MIB variables
+            returned in SNMP response.
+
+        """
+        (error_status, error_index, var_binds) = args
+
+        if error_status:
+            self.log.error('%s: %s at %s' % (self.address,
+                                             error_status.prettyPrint(),
+                                             error_index and
+                                             var_binds[int(error_index) - 1][0] or '?'))
+        else:
+            for var in var_binds:
+                self.log.debug(' = '.join([x.prettyPrint() for x in var]))
+
+        return var_binds
+
+    def _failure(self, error_indication):
+        """Failure Errback for getCmd.
+
+        Taken from Twisted example for SNMPv1 from pySNMP documentation:
+        http://snmplabs.com/pysnmp/examples/hlapi/twisted/contents.html
+
+        """
+        self.log.error('%s failure: %s' % (self.address, error_indication))
+
+    def get(self, oid_list):
+        """Issue a getCmd to get SNMP OID states.
```

Comment:
This docstring could be a touch more helpful... although it gives examples of calling get(), it does not tell me what get() returns, nor more importantly how I can extract the gotten values.

Perhaps be a little more pointed on this docstring (what, exactly, does get() return?  A Deferred, which will callback with the var_binds list from self._success...) and then show a more complete usage example in the class docstring.

Reply:
Attempted to make more clear in 0be0a90. The returned objects from getCmd I find awkward to work with. I'm reaching into a hidden value `._value` to get the integer the corresponds to the description string in the Enum, as there doesn't seem to be a good way to grab it otherwise, just the actual description string (which I did start saving to the .g3 files as well in b8676d5.) They seemed well designed to print, as seen in the example.